src/main/java/com/puppycrawl/tools/checkstyle/checks/regexp/RegexpCheck.java

```diff
@@ -304,69 +295,59 @@ public int[] getRequiredTokens() {
     @SuppressWarnings("deprecation")
     @Override
     public void beginTree(DetailAST rootAST) {
-        matcher = format.matcher(getFileContents().getText().getFullText());
-        matchCount = 0;
-        errorCount = 0;
-        findMatch();
+        final Matcher matcher = format.matcher(getFileContents().getText().getFullText());
+        processRegexpMatches(matcher);
     }
 
     /**
-     * Recursive method that finds the matches.
+     * Finds the matches in the file.
      *
-     * @noinspection TailRecursion
-     * @noinspectionreason TailRecursion - until issue #14814
+     * @param matcher The matcher
      */
     // suppress deprecation until https://github.com/checkstyle/checkstyle/issues/11166
+    // -@cs[CyclomaticComplexity] Most simplified logic
     @SuppressWarnings("deprecation")
-    private void findMatch() {
-        final boolean foundMatch = matcher.find();
-        if (foundMatch) {
+    private void processRegexpMatches(Matcher matcher) {
+        // Tracks number of errors.
+        int errorCount = 0;
+        int matchCount = 0;
+        while (errorCount <= errorLimit - 1 && matcher.find()) {
             final FileText text = getFileContents().getText();
             final LineColumn start = text.lineColumn(matcher.start());
             final int startLine = start.getLine();
 
-            final boolean ignore = isIgnore(startLine, text, start);
-
+            final boolean ignore = isIgnore(startLine, text, start, matcher);
             if (!ignore) {
                 matchCount++;
                 if (illegalPattern || checkForDuplicates
                         && matchCount - 1 > duplicateLimit) {
                     errorCount++;
-                    logMessage(startLine);
+                    logMessage(startLine, errorCount);
+                }
+                if (!illegalPattern && !checkForDuplicates) {
+                    break;
```

Comment:
Could you explain please why do we need this condition? Can we remove it?

Reply:
this is simplified logic version of what ```canContinueValidation``` was doing previously 
```
   private boolean canContinueValidation(boolean ignore) {
        return errorCount <= errorLimit - 1
                && (ignore || illegalPattern || checkForDuplicates);
    }
```
takes care of second condition like this : ```!illegalPattern && !checkForDuplicates && !ignore```

Reply:
Did you try to remove this condition? What was the result?

Reply:
this condition is only way to break out the loop. tests start to fail if we remove the entire condition or just keep the ```break``` statement

Reply:
we can't put ```!illegalPattern && !checkForDuplicates``` in the loop's condition either because ignore needs to be false at the same time

Reply:
Ok . I got it . yes you are right . this condition is indeed redundant. I must have not been checking it properly.